### PR TITLE
Fix inconsistent exception type in response.json() method

### LINKED ISSUE #5794: response.json() raises inconsistent exception type
There are some comments about this on https://github.com/psf/requests/issues/4842 but I think it warrants its own issue.

If `simplejson` is present in the environment, the library uses it so `.json()` returns `simplejson.errors.JSONDecodeError` rather than `json.decoder.JSONDecodeError`

If I'm writing a library that uses `requests` I don't know what other libraries will be present in an environment. I expect that a method returns a consistent exception type that I can handle. As it stands, anyone writing a library (e.g. https://github.com/conda-forge/conda-smithy/pull/1369, https://github.com/Vonage/vonage-python-sdk/issues/197) has to check themselves if `simplejson` is present to know what exception type will be thrown.

### PR DESCRIPTION
**Pull request by steveberdy**
## Summary
Fixes the inconsistency of errors thrown in the `response.json()` method in `models.py`, while preserving backwards compatibility. 

### What we used to have
Depending on whether or not `simplejson` was installed in the user's library, and whether the user was running Python 3+ versus Python 2, a different exception would get thrown when there was an issue with decoding response text as JSON. If `simplejson` was installed, `simplejson.JSONDecodeError` would get thrown. If not, `json.JSONDecodeError` would get thrown for Python 3+ users, and `ValueError` for Python 2 users. Thus, depending on the scenario, users would find themselves having to catch either `simplejson.JSONDecodeError`, `json.JSONDecodeError`, or `ValueError` in the case of the `response.json()` method failing. This inconsistency is not ideal.

### What we have now
There is now one error class in `exceptions.py` that will represent all errors that can be thrown or caught in the `response.json()` method: `requests.JSONDecodeError`. All `simplejson` functionality was replaced with `json` functionality, but the new error type aliases both `json.JSONDecodeError` and `simplejson.JSONDecodeError`. If `simplejson` is not installed, its `JSONDecodError` is replaced with a plain `Exception` when aliased. If the user is running Python 3+, `json.JSONDecodeError` is aliased, but if the user is running Python 2, that is replaced with `ValueError` when aliased, as the `JSONDecodeError` was not previously a part of the `json` library. Now, all five error types, `json.JSONDecodeError`, `simplejson.JSONDecodeError`, `ValueError`, and `requests.JSONDecodeError` and its parent class `requests.RequestException`, will be caught from the `requests.JSONDecodeError` that will be raised.

Fixes #5794
@sigmavirus24

### CODE REVIEW

File: requests/compat.py [hunk 1]
```diff
@@ -25,10 +26,6 @@
 #: Python 3.x?
 is_py3 = (_ver[0] == 3)
 
-try:
-    import simplejson as json
-except ImportError:
-    import json
```
Comment (sigmavirus24) [inline requests/compat.py]: This is backwards incompatible and needs to be reverted

File: requests/exceptions.py [hunk 2]
```diff
@@ -25,6 +35,10 @@ def __init__(self, *args, **kwargs):
         super(RequestException, self).__init__(*args, **kwargs)
 
 
+class JSONDecodeError(StandardJSONDecodeError, SimpleJSONDecodeError):
+    """Couldn't decode the text into json"""
+
+
 class InvalidJSONError(RequestException):
```
Comment (sigmavirus24) [inline requests/exceptions.py]: I forgot we had this exception, where do we use it? Should we sub-class this in `JSONDecodeError` as well?

File: requests/models.py [hunk 3]
```diff
@@ -39,7 +40,7 @@
     Callable, Mapping,
     cookielib, urlunparse, urlsplit, urlencode, str, bytes,
     is_py2, chardet, builtin_str, basestring)
-from .compat import json as complexjson
+from .compat import json
```
Comment (sigmavirus24) [inline requests/models.py]: Why did you change this?

File: requests/models.py [hunk 4]
```diff
@@ -176,12 +177,14 @@ def register_hook(self, event, hook):
         """Properly register a hook."""
 
         if event not in self.hooks:
-            raise ValueError('Unsupported event specified, with event name "%s"' % (event))
+            raise ValueError(
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this. It's tangential to the change

File: requests/models.py [hunk 5]
```diff
@@ -176,12 +177,14 @@ def register_hook(self, event, hook):
...
+                'Unsupported event specified, with event name "%s"' % (event))
 
         if isinstance(hook, Callable):
             self.hooks[event].append(hook)
         elif hasattr(hook, '__iter__'):
-            self.hooks[event].extend(h for h in hook if isinstance(h, Callable))
+            self.hooks[event].extend(
+                h for h in hook if isinstance(h, Callable))
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this as well

File: requests/models.py [hunk 6]
```diff
@@ -224,8 +227,8 @@ class Request(RequestHooksMixin):
     """
 
     def __init__(self,
-            method=None, url=None, headers=None, files=None, data=None,
-            params=None, auth=None, cookies=None, hooks=None, json=None):
+                 method=None, url=None, headers=None, files=None, data=None,
+                 params=None, auth=None, cookies=None, hooks=None, json=None):
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this

File: requests/models.py [hunk 7]
```diff
@@ -308,8 +311,8 @@ def __init__(self):
         self._body_position = None
 
     def prepare(self,
-            method=None, url=None, headers=None, files=None, data=None,
-            params=None, auth=None, cookies=None, hooks=None, json=None):
+                method=None, url=None, headers=None, files=None, data=None,
+                params=None, auth=None, cookies=None, hooks=None, json=None):
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this

File: requests/models.py [hunk 8]
```diff
@@ -384,7 +387,8 @@ def prepare_url(self, url, params):
             raise InvalidURL(*e.args)
 
         if not scheme:
-            error = ("Invalid URL {0!r}: No schema supplied. Perhaps you meant http://{0}?")
+            error = (
+                "Invalid URL {0!r}: No schema supplied. Perhaps you meant http://{0}?")
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this

File: requests/models.py [hunk 9]
```diff
@@ -438,7 +442,8 @@ def prepare_url(self, url, params):
             else:
                 query = enc_params
 
-        url = requote_uri(urlunparse([scheme, netloc, path, None, query, fragment]))
+        url = requote_uri(urlunparse(
+            [scheme, netloc, path, None, query, fragment]))
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this

File: requests/models.py [hunk 10]
```diff
@@ -500,7 +505,8 @@ def prepare_body(self, data, files, json=None):
                     self._body_position = object()
 
             if files:
-                raise NotImplementedError('Streamed bodies and files are mutually exclusive.')
+                raise NotImplementedError(
+                    'Streamed bodies and files are mutually exclusive.')
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this

File: requests/models.py [hunk 11]
```diff
@@ -776,7 +782,8 @@ def generate():
         if self._content_consumed and isinstance(self._content, bool):
             raise StreamConsumedError()
         elif chunk_size is not None and not isinstance(chunk_size, int):
-            raise TypeError("chunk_size must be an int, it is instead a %s." % type(chunk_size))
+            raise TypeError(
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this

File: requests/models.py (hunk 10 above)
Comment (steveberdy) [reply requests/models.py:509]: Ok

File: requests/models.py [hunk 12]
```diff
@@ -833,7 +840,8 @@ def content(self):
             if self.status_code == 0 or self.raw is None:
                 self._content = None
             else:
-                self._content = b''.join(self.iter_content(CONTENT_CHUNK_SIZE)) or b''
+                self._content = b''.join(
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this

File: requests/models.py (hunk 10 above)
Comment (steveberdy) [reply requests/models.py:509]: I think that was auto-formatted

File: requests/models.py [hunk 13]
```diff
@@ -882,12 +890,8 @@ def json(self, **kwargs):
         r"""Returns the json-encoded content of a response, if any.
 
         :param \*\*kwargs: Optional arguments that ``json.loads`` takes.
-        :raises simplejson.JSONDecodeError: If the response body does not
-            contain valid json and simplejson is installed.
-        :raises json.JSONDecodeError: If the response body does not contain
-            valid json and simplejson is not installed on Python 3.
-        :raises ValueError: If the response body does not contain valid
-            json and simplejson is not installed on Python 2.        
+        :raises requests.JSONDecodeError: If the response body does not
```
Comment (sigmavirus24) [inline requests/models.py]: Please document this as `requests.exceptions.JSONDecodeError`. Eventually I desperately want to stop importing every exception class into `requests/__init__.py`

Comment (steveberdy) [reply requests/models.py:893]: Alright!

File: requests/models.py [hunk 14]
```diff
@@ -907,7 +911,13 @@ def json(self, **kwargs):
...
                     # used.
                     pass
-        return complexjson.loads(self.text, **kwargs)
+
+        try:
+            return json.loads(self.text, **kwargs)
+        except json.JSONDecodeError as e:
+            # Catch JSON-related errors and raise as requests.JSONDecodeError
+            # This aliases json.JSONDecodeError and simplejson.JSONDecodeError
+            raise JSONDecodeError(e.msg, e.doc, e.pos)
```
Comment (sigmavirus24) [inline requests/models.py]: We should probably preserve the original exception instance in this as well and proxy attributes down to it. That way folks can find the original exception. I'm not certain `simplejson` and `json` have the same exception attributes with the same meaning

File: requests/models.py [hunk 15]
```diff
@@ -944,10 +954,12 @@ def raise_for_status(self):
             reason = self.reason
 
         if 400 <= self.status_code < 500:
-            http_error_msg = u'%s Client Error: %s for url: %s' % (self.status_code, reason, self.url)
+            http_error_msg = u'%s Client Error: %s for url: %s' % (
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this

File: requests/models.py [hunk 16]
```diff
@@ -944,10 +954,12 @@ def raise_for_status(self):
...
+                self.status_code, reason, self.url)
 
         elif 500 <= self.status_code < 600:
-            http_error_msg = u'%s Server Error: %s for url: %s' % (self.status_code, reason, self.url)
+            http_error_msg = u'%s Server Error: %s for url: %s' % (
```
Comment (sigmavirus24) [inline requests/models.py]: Please revert this

File: docs/user/quickstart.rst [hunk 17]
```diff
@@ -150,9 +150,7 @@ There's also a builtin JSON decoder, in case you're dealing with JSON data::
 
 In case the JSON decoding fails, ``r.json()`` raises an exception. For example, if
 the response gets a 204 (No Content), or if the response contains invalid JSON,
-attempting ``r.json()`` raises ``simplejson.JSONDecodeError`` if simplejson is
-installed or raises ``ValueError: No JSON object could be decoded`` on Python 2 or
-``json.JSONDecodeError`` on Python 3.
+attempting ``r.json()`` raises ``requests.JSONDecodeError``.
```
Comment (sigmavirus24) [inline docs/user/quickstart.rst]: We must be clear that this is _new_ in the next version of Requests (add it to our changelog too) and that it is backwards compatible as it inherits from the exceptions raised previously

Comment (steveberdy) [reply docs/user/quickstart.rst:153]: Ok, I'll indicate that.

Comment (steveberdy) [reply docs/user/quickstart.rst]: Do you want me to add it to the changelog now or do you like to do that?

File: requests/models.py (hunk 3 above)
Comment (steveberdy) [reply requests/models.py:43]: I'll change it back. It seemed that just `json` would reflect the changes made.

File: requests/exceptions.py (hunk 2 above)
Comment (steveberdy) [reply requests/exceptions.py]: I added this exception to sub-class both `JSONDecodeError`s

File: docs/user/quickstart.rst (hunk 17 above)
Comment (sigmavirus24) [reply docs/user/quickstart.rst]: Please add it to the changelog now

Comment (steveberdy) [reply docs/user/quickstart.rst:153]: Ok, will do!

File: requests/models.py (hunk 14 above)
Comment (steveberdy) [reply requests/models.py:913]: I could catch a `ValueError` and that will get caught for both `simplejson` and `json` exceptions. Sound good?

File: docs/user/quickstart.rst (hunk 17 above)
Comment (steveberdy) [reply docs/user/quickstart.rst]: Include it under the `dev` category?

File: requests/models.py (hunk 14 above)
Comment (steveberdy) [reply requests/models.py:913]: @sigmavirus24  Actually, at least in the latest versions of both `json` and `simplejson`, they both have `msg`, `doc`, and `pos` attributes. I'll add an extra `except` for `simplejson`.

Comment (steveberdy) [reply requests/models.py:913]: Both will get caught as `json.JSONDecodeError` since either are retrieved from `compat.py` as `json`.

**Comment by steveberdy** [comment]
@sigmavirus24 I made the changes you asked for. The PR is ready for review again.

File: requests/exceptions.py (hunk 2 above)
Comment (sigmavirus24) [reply requests/exceptions.py]: I'm talking about the `InvalidJSONError` exception here (the line I commented on)

**Review by sigmavirus24 — APPROVED** [review]

**Comment by steveberdy** [comment]
Looks like the only problem is with my test file. We can remove that. Then are we safe to merge?

**Comment by steveberdy** [comment]
Can we run the workflows?

**Comment by steveberdy** [comment]
When I did the minor changes I didn't change all instances in `models.py` of `json` with `complexjson`. Now that has been completed. That should fix the problem if we were to run the workflows again.

**Comment by steveberdy** [comment]
I ran multiple tests and it looks good. Can we repeat the workflows?

**Comment by steveberdy** [comment]
@sigmavirus24 can we repeat the workflows? I made the necessary changes.

File: docs/user/quickstart.rst [hunk 18]
```diff
@@ -153,9 +153,9 @@ There's also a builtin JSON decoder, in case you're dealing with JSON data::
 
 In case the JSON decoding fails, ``r.json()`` raises an exception. For example, if
 the response gets a 204 (No Content), or if the response contains invalid JSON,
-attempting ``r.json()`` raises ``simplejson.JSONDecodeError`` if simplejson is
-installed or raises ``ValueError: No JSON object could be decoded`` on Python 2 or
-``json.JSONDecodeError`` on Python 3.
+attempting ``r.json()`` raises ``requests.JSONDecodeError``. This is new for the
```
Comment (nateprewitt) [inline docs/user/quickstart.rst]: We probably don't want to add all the information in the last sentence as it will become stale quickly. Let's simplfy it to something closer to:

> This wrapper exception provides interoperability for multiple exceptions that may be thrown by different python versions and json serialization libraries.

File: requests/compat.py [hunk 19]
```diff
@@ -33,6 +33,7 @@
 except ImportError:
     import json
 
+
```
Comment (nateprewitt) [inline requests/compat.py]: These spacing changes aren't really related to the current PR, can we remove them.

File: requests/exceptions.py [hunk 20]
```diff
@@ -25,6 +35,10 @@ def __init__(self, *args, **kwargs):
         super(RequestException, self).__init__(*args, **kwargs)
 
 
+class JSONDecodeError(StandardJSONDecodeError, SimpleJSONDecodeError):
```
Comment (nateprewitt) [inline requests/exceptions.py]: Our custom exceptions in Requests should ideally fall under a catch-all parent exception, `RequestException`. That allows users to catch everything thrown by Requests.

File: requests/exceptions.py (hunk 2 above)
Comment (nateprewitt) [reply requests/exceptions.py]: I think we're still waiting on a response here, @steveberdy.

**Comment by steveberdy** [comment]
@sigmavirus24 it looks like all the builds succeeded for Python 3+. Does that make this safe to merge? I don't know why Python 2 is having a problem with the module; I'm not sure if that's legitimate.

File: requests/models.py (hunk 14 above)
Comment (nateprewitt) [reply requests/models.py]: I'm not sure I'm following how this will work in the Python 2 `json` case. Where is `ValueError` being handled?

**Review by nateprewitt — CHANGES_REQUESTED** [review]

File: requests/models.py (hunk 14 above)
Comment (nateprewitt) [reply requests/models.py]: https://github.com/psf/requests/pull/5856/checks?check_run_id=3031103164 Yeah, this is broken. We need to be able to handle this a different way.

File: docs/user/quickstart.rst (hunk 18 above)
Comment (steveberdy) [reply docs/user/quickstart.rst:156]: Ok. I'll change that.

File: requests/compat.py (hunk 19 above)
Comment (steveberdy) [reply requests/compat.py:36]: I'll remove that.

File: requests/exceptions.py (hunk 20 above)
Comment (steveberdy) [reply requests/exceptions.py:38]: The point of this exception was to catch other `JSONDecodError`s. A `RequestException` would imply something is wrong with the request, when the only thing that went wrong is the json decoding process. This is to indicate that the response text was not json, not that the request itself failed.

**Comment by steveberdy** [comment]
@nateprewitt for Python 2, `ValueError` is thrown rather than `json.JSONDecodeError`. For Python 3+, the `JSONDecodeError`s inherit `ValueError`.

File: requests/models.py (hunk 14 above)
Comment (nateprewitt) [reply requests/models.py]: I see your last comment was made outside of this thread, @steveberdy. To clarify, while `json.JSONDecodeError` does inherit from `ValueError`, it's not going to work in this case. The exception doesn't exist in Python 2, so we fall over with an attribute error when handling the exception. You'll need to take another approach to handle this.

File: requests/compat.py (hunk 19 above)
Comment (steveberdy) [reply requests/compat.py]: @nateprewitt I made the two above changes in my most recent commit

**Comment by steveberdy** [comment]
Same issues with Python 2 again. Is that normal?

**Comment by nateprewitt** [comment]
@steveberdy Please refer to my previous comments in the code review. Your current approach isn't going to work in Python 2. I'd recommend using your exception resolution in compat.py to handle this. I think there are also some outstanding comments that are still unresolved.

File: requests/exceptions.py (hunk 20 above)
Comment (steveberdy) [reply requests/exceptions.py:38]: I made `JSONDecodeError` a subclass of `RequestException` as well.

**Comment by steveberdy** [comment]
I'm not sure how I can use `compat.py` to make the `json` reference Python 2 compatible, since I don't know what makes Python 2 throw the module error, even after doing some research. Any advice? Also, I got a method resolution order (MRO) error for my last commit, which I never got while testing, so I changed the order of super-classes in `JSONDecodeError` from
```python
class JSONDecodeError(StandardJSONDecodeError, SimpleJSONDecodeError, RequestException)
```
to 
```python
class JSONDecodeError(RequestException, StandardJSONDecodeError, SimpleJSONDecodeError)
```
Any advice on this as well? I appreciate your help.

File: requests/exceptions.py (hunk 2 above)
Comment (steveberdy) [reply requests/exceptions.py]: Yes, I can subclass that in `JSONDecodeError` as well.

File: requests/models.py (hunk 14 above)
Comment (steveberdy) [reply requests/models.py:913]: Hi @nateprewitt. I'm sorry if this has been frustrating -- I've just found this comment. I'll be working on this in `compat.py`. Do you know of any good ways to test a Python 2 environment when I have Python 3 installed?

**Comment by steveberdy** [comment]
@sigmavirus24 @nateprewitt Please check my latest changes and run any workflows to test that they work in all environments. I made edits to `compat.py`, `models.py`, and `exceptions.py`, as well as changed some documentation accordingly.

**Comment by steveberdy** [comment]
Is this a proper fix for the Python 2 errors?
```python
try:
    return complexjson.loads(self.text, **kwargs)
except JSONDecodeError as e:
    # Catch JSON-related errors and raise as requests.JSONDecodeError
    # This aliases json.JSONDecodeError and simplejson.JSONDecodeError
    if is_py2: # e is a ValueError
        raise RequestsJSONDecodeError()
    else:
        raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
```

**Comment by steveberdy** [comment]
@sigmavirus24 @nateprewitt I made the necessary edits to fix workflow failures. Any feedback on Python 2 compatibility before running the workflows again? Thanks.

**Comment by nateprewitt** [comment]
Hi @steveberdy, I'd recommend setting up a testing environment locally with our tox script and something like `pyenv` to manage Python versions. If you're having difficulty with that, you can test changes yourself on another branch using Github actions in your own repository. That will help expedite your testing cycle and avoid needing to ping this thread every time a change is made.

**Comment by steveberdy** [comment]
@nateprewitt Thank you for the advice. I ran the workflows, and all tests passed: https://github.com/steveberdy/requests/actions/runs/1023889042

**Comment by steveberdy** [comment]
Can you run your workflows so this can be ready to merge?

**Comment by steveberdy** [comment]
Is this ready to merge?

File: requests/models.py [hunk 21]
```diff
@@ -907,7 +905,16 @@ def json(self, **kwargs):
...
+
+        try:
+            return complexjson.loads(self.text, **kwargs)
+        except JSONDecodeError as e:
+            # Catch JSON-related errors and raise as requests.JSONDecodeError
+            # This aliases json.JSONDecodeError and simplejson.JSONDecodeError
+            if is_py2: # e is a ValueError
+                raise RequestsJSONDecodeError()
+            else:
+                raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
```
Comment (sigmavirus24) [inline requests/models.py]: Giving the user 0 useful information is unacceptable. Why can't we use `e.message` there?

```py
Python 2.7.18 (default, May 19 2021, 00:00:00)
[GCC 11.1.1 20210428 (Red Hat 11.1.1-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import json
>>> try:
...   json.loads("foo")
... except ValueError as ve:
...   pass
...
>>> dir(ve)
['__class__', '__delattr__', '__dict__', '__doc__', '__format__', '__getattribute__', '__getitem__', '__getslice__', '__hash__', '__init__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '__unicode__', 'args', 'message']
>>> ve.message
'No JSON object could be decoded'
>>> ve.args
('No JSON object could be decoded',)
>>>
```

Comment (steveberdy) [reply requests/models.py]: I added that. Workflows passed. Thanks for mentioning this!

File: requests/compat.py [hunk 22]
```diff
@@ -28,8 +28,10 @@
 #: Python 3.x?
 is_py3 = (_ver[0] == 3)
 
+has_simplejson = False
```
Comment (nateprewitt) [inline requests/compat.py:31]: It doesn't look like we use this anywhere except to declare the exceptions. We're already setting either dependency to be imported under the `json` namespace. Can we not just do `from json import JSONDecodeError` under the py3 section and have it automatically resolve? Perhaps adding a short comment above the import saying it's using simplejson or json depending on what was resolved?

File: HISTORY.md [hunk 23]
```diff
@@ -4,13 +4,16 @@ Release History
 dev
 ---
 
--   \[Short description of non-trivial change.\]
-
```
Comment (sigmavirus24) [inline HISTORY.md]: Why did you remove this?

File: HISTORY.md [hunk 24]
```diff
@@ -4,13 +4,16 @@ Release History
...
 2.26.0 (2021-07-13)
 -------------------
 
 **Improvements**
 
+- Added a `requests.exceptions.JSONDecodeError` to decrease inconsistencies
+  in the library. This gets raised in the `response.json()` method, and is
+  backwards compatible as it inherits from previously thrown exceptions.
+  Can be caught from `requests.exceptions.RequestException` as well.
+
```
Comment (sigmavirus24) [inline HISTORY.md]: 2.26.0 was released. This can not be added to those notes as this was not released with 2.26.0 please move it above the `2.26.0 (2021-07-13)` line.

File: docs/api.rst [hunk 25]
```diff
@@ -127,7 +127,7 @@ API Changes
   ::
 
       import requests
-      r = requests.get('https://github.com/timeline.json')
+      r = requests.get('https://api.github.com/events')
```
Comment (sigmavirus24) [inline docs/api.rst]: These changes in this file are unrelated and should be in a separate PR

**Review by sigmavirus24 — CHANGES_REQUESTED** [review]

File: HISTORY.md (hunk 23 above)
Comment (steveberdy) [reply HISTORY.md:8]: There were merge conflicts, and the remote already had that overwritten.

**Review by sigmavirus24 — APPROVED** [review]

File: requests/exceptions.py [hunk 26]
```diff
@@ -8,6 +8,16 @@
 """
 from urllib3.exceptions import HTTPError as BaseHTTPError
 
+try:
```
Comment (nateprewitt) [inline requests/exceptions.py]: We have logic for which simplejson/json exceptions we're using spread across a few places now. Is there a compelling reason not to simply import JSONDecodeError from compat since it's already been resolved and supplying it as a single parent class?

**Review by nateprewitt — COMMENTED** [review]
Thanks for the updates, @steveberdy. I added a couple of comments for discussion. I'm not convinced they're necessarily the right direction, but it does feel like the simplejson/json problems are potentially being spread more places than needed.

Let me know what you think, thanks!

File: requests/exceptions.py (hunk 26 above)
Comment (steveberdy) [reply requests/exceptions.py]: @nateprewitt We want to know which properties the `JSONDecodeError` has, since it may have different properties depending on compatibility. That requires an extra check on import.

Comment (nateprewitt) [reply requests/exceptions.py]: @steveberdy I'm not sure I'm following, could you clarify which properties you're referring to? We have 3 parent classes that we'd ever use in our environment, `json.JSONDecodeError`, `simplejson.JSONDecodeError`, and `ValueError`. Only one of those can be relevant in a given execution of Requests.

What I was asking about was if we did:

```python
from .compat import JSONDecodeError as CompatJSONDecodeError

[...]

class JSONDecodeError(InvalidJSONError, CompatJSONDecodeError):
```

We've already resolved these values once in compat, so I wanted to know why we're doing it again here.

Comment (steveberdy) [reply requests/exceptions.py]: @nateprewitt For Python 3, in the `compat.py` file itself, reference to the `json` package directly, whether aliased as `json` or actually the `json` package itself, would always look for the `json` package rather than whatever was aliased as it. So, just in the `compat.py` file itself, I used the check. From all other scripts that have `from .compat import json`, however, it works as expected.

Comment (steveberdy) [reply requests/exceptions.py]: Did you have any more questions about the code?

Comment (nateprewitt) [reply requests/exceptions.py]: Hey @steveberdy, I think there may still be some confusion. If you open the _Files changed_ tab, you should be able to see the full context of this thread. I've written out the [full diff](https://github.com/nateprewitt/requests/commit/665cf30861fb356d84103808e240d6940e8ff3d6) from the suggestion above to minimize ambiguity. You should find this passes the test suite without issue, so please let me know if there were other concerns you had.

While we're here it would also be good to add a test exercising this change. Something to the effect of the example below should be sufficient:

```
def test_json_decode_compatibility(self, httpbin):
    r = requests.get(httpbin('bytes/20'))
    with pytest.raises(requests.exceptions.JSONDecodeError):
        r.json()
```

This won't cover simplejson in our general test suite, but I've done a one off run to ensure that's working.

Comment (steveberdy) [reply requests/exceptions.py]: Thank you, that makes total sense. I made those changes and I'm running workflows right now.

**Review by sigmavirus24 — APPROVED** [review]

**Review by nateprewitt — APPROVED** [review]
Cool, I think this looks like we’re ready now. Thanks for seeing this through, @steveberdy.

For merging, I’m not sure if we want to put this on a 2.27.x branch while we wait for the next release, or if we merge now knowing we won’t do another patch release for 2.26.x. Any thoughts @sigmavirus24?

**Comment by sigmavirus24** [comment]
Let's just commit to the next release being 2.27.0. Heck we can push that out whenever and if we can get the proxy stuff merged too (once it's in a good state) then I think we can cut 2.27 with just these two "features"

**Comment by nateprewitt** [comment]
Alright, sounds like a plan to me. Thanks again, @steveberdy.
